pkg/cmd/server/kubernetes/node_config.go

```diff
@@ -102,7 +102,15 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 	server := kapp.NewKubeletServer()
 	server.Config = path
 	server.RootDirectory = options.VolumeDirectory
-	server.HostnameOverride = options.NodeName
+
+	// kubelet finds the node IP address by doing net.ParseIP(hostname) and if that fails,
+	// it does net.LookupIP(NodeName) and picks the first non-loopback address.
+	// Pass node IP as hostname to make kubelet use the desired IP address.
+	if len(options.NodeIP) > 0 {
+		server.HostnameOverride = options.NodeIP
```

Comment:
Should we also check for loopback here? If we do not then cli may end up showing loopback for hostnames (which may clash) because status update will pick the hostnameoverride anyway whether the IPaddress is properly calculated or not.
Not critical.